assigner/core/assigner.go

```diff
@@ -31,22 +32,39 @@ type Assigner struct {
 	p2pHost host.Host
 	// policy decides what publisher to accept announce messages from.
 	policy peerutil.Policy
+	// presets maps publisher ID to pre-assigned indexers.
+	presets map[peer.ID][]int
 	// receiver receives announce messages.
 	receiver *announce.Receiver
 	// replication is the number of indexers to assign a publisher to.
 	replication int
 	// watchDone signals that the watch function exited.
 	watchDone chan struct{}
 	// waitingNotice are channels waiting for a specific peer to be assigned
-	waitingNotice map[peer.ID]chan string
+	waitingNotice map[peer.ID]chan int
 	noticeMutex   sync.Mutex
 }
 
 type assignment struct {
-	indexers   map[int]struct{}
+	indexers   []int
 	processing bool
 }
 
+func (asmt *assignment) addIndexer(x int) {
+	i := sort.SearchInts(asmt.indexers, x)
+	if i < len(asmt.indexers) && asmt.indexers[i] == x {
+		return
+	}
+	asmt.indexers = append(asmt.indexers, 0)
+	copy(asmt.indexers[i+1:], asmt.indexers[i:])
```

Comment:
it is non-obvious to me that this is safe, and that it doesn't cause the value at i to be 'smeared' over the rest of the array.

Reply:
This is a typical insert into a slice:
https://gist.github.com/GimmyHchs/33bd06e68d72a913a8587b09d41b50d0#insert